devdoc/critical_section_requirements.md

```diff
@@ -0,0 +1,47 @@
+`critical_section` requirements
+================
+
+## Overview
+
+`critical_section` is module that provides a few helper functions that let us use a `volatile_atomic int32_t` as a thread synchronization object with `enter_crit_section` and `leave_crit_section` functionality.
+
+It differs from c-pal's `srw_lock` in 2 important respects:
+1. It does not require any `_create` or `_destroy` calls.  It only requires a `volatile_atomic int32_t` which is much simpler to allocate.
+2. All access is exclusive.  There is no concept of shared access.
+
+## Exposed API
+
+```c
+    MOCKABLE_FUNCTION(, void, enter_crit_section, volatile_atomic int32_t*, access_value);
+    MOCKABLE_FUNCTION(, void, leave_crit_section, volatile_atomic int32_t*, access_value);
+```
+
+### enter_crit_section
+
+```c
+    MOCKABLE_FUNCTION(, void, enter_crit_section, volatile_atomic int32_t*, access_value);
+```
+`enter_crit_section` blocks until it is safe to enter the critical section.
+
+**SRS_CRITICAL_SECTION_18_001: [** If `access_value` is NULL, `enter_crit_section` shall terminate the process. **]**

```

Comment:
Hmmmm, we usually don't do this. It is highly arguable, but generally arg checks will just error out for us

Reply:
I was too lazy to add a return, but your wisdom once again shines. Fixed.